docs: add standard deprovisioning logic into the docs (#3757)
aws/karpenter: website/content/en/preview/concepts/deprovisioning.md

The finalizer blocks deletion of the node object while the Termination Controller cordons and drains the node, before removing the underlying machine. Deprovisioning is triggered by the Deprovisioning Controler, by the user through manual deprovisioning, or through an external system that sends a delete request to the node object.

### Deprovisioning Controller
Karpenter automatically discovers deprovisionable nodes and spins up replacements when needed. Karpenter deprovisions nodes by executing one [automatic method](#methods) at a time, in order of Expiration, Drift, Emptiness, then Consolidation. Each method varies slightly but they all follow the standard deprovisioning process:
1. Identify a list of prioritized candidates for the deprovisioning method.
   1. If there are no deprovisionable nodes, continue to the next deprovisioning method.
2. For each deprovisionable node, execute a scheduling simulation with the pods on the node to find if any replacement nodes are needed.
3. Cordon the node(s) to prevent pods from scheduling to it.
4. Pre-spin any replacement nodes needed as calculated in Step (2), and wait for them to become ready.

>    1. If a replacement node fails to initialize, un-cordon the node(s), and restart from at Step (1).

Reviewer: ```suggestion
   1. If a replacement node fails to initialize, un-cordon the node(s), and restart from Step (1).
```

nit

5. Delete the node(s) and wait for the termination controller to gracefully shutdown the node(s).

> 6. Once the termination controller terminates the node, go back to Step (1), starting at the the first method again.

Reviewer: ```suggestion
6. Once the termination controller terminates the node, go back to Step (1), starting at the the first deprovisoining method again.
```

nit: maybe just to make this a bit more clear

### Termination Controller
When a Karpenter node is deleted, the Karpenter finalizer will block deletion and the APIServer will set the `DeletionTimestamp` on the node, allowing Karpenter to gracefully shutdown the node, modeled after [K8s Graceful Node Shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown). Karpenter's graceful shutdown process will:
1. Cordon the node to prevent pods from scheduling to it.
2. Begin evicting the pods on the node with the [K8s Eviction API](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/) to respect PDBs, while ignoring all non-daemonset pods and [static pods](https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/). Wait for the node to be fully drained before proceeding to Step (3).

>    1. While waiting, if the underlying instance of the node no longer exists, remove the finalizer to allow the APIServer to delete the node, completing termination.

Reviewer: ```suggestion
   1. While waiting, if the underlying machine for the node no longer exists, remove the finalizer to allow the APIServer to delete the node, completing termination.
```

nit: probably want to start using this lingo and being consistent throughout our docs to get users familiar with the naming

---

Karpenter sets a Kubernetes [finalizer](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) on each node it provisions.

> The finalizer signals to the Karpenter Termination controller to gracefully shutdown a node that receives a deletion request. Karpenter deprovisioning can be executed by Karpenter through the Deprovisioning Controler or by the user through manual deprovisioning.

Reviewer 2: ```suggestion
The finalizer blocks deletion of the node object while the termination controller cordons and drains the node, before removing the underlying machine. Deprovisioning is triggered by deprovisioning controller or by a user 
 or external system that sends a DELETE request to the node object.
```

Author: Took some bits and added it in. Let me know what you think.

> ### Deprovisioning Controller

Author: Ah ok, I changed it to add the ## Control Flow at the top so this should be fixed.

> Karpenter automatically discovers deprovisionable nodes and spins up replacements when needed. Karpenter deprovisions nodes by executing one [automatic method](#methods) at a time, in order of Expiration, Drift, Emptiness, then Consolidation. Each method varies slightly but they all follow the standard deprovisioning process:

Reviewer 2: ```suggestion
Karpenter automatically discovers deprovisionable nodes and spins up replacements when needed. Karpenter deprovisions nodes by executing one [method](#methods) at a time, in order of Expiration, Drift, Emptiness, and Consolidation. Each method varies slightly but they all follow the standard deprovisioning process:
```

Author: I chose to delineate like this to be clear and follow descriptions below, and to indicate that there is an order. Can you expand more here?

> Karpenter automatically discovers deprovisionable nodes and spins up replacements when needed. Karpenter deprovisions nodes by executing one [automatic method](#methods) at a time, in order of Expiration, Drift, Emptiness, then Consolidation. Each method varies slightly but they all follow the standard deprovisioning process:

Author: The methods are called out below and then expanded on even further. I don't include interruption here because it only follows the termination controller, not the deprovisioning controller.

> 1. Generate a list of deprovisionable nodes and sort them as needed.

Reviewer 2: ```suggestion
1. Identify a prioritized list of candidates for deprovisioning.
```

   1. If there are no deprovisionable nodes, continue to the next deprovisioning method.
2. For each deprovisionable node, execute a scheduling simulation with the pods on the node to find if any replacement nodes are needed.

> 3. Cordon to-be-deprovisioned node(s) so no further pods can be scheduled on the node.

Reviewer: You're using `to-be-deprovisioned` a lot and, to me, it feels hard to read. Can we use something shorter or change this to be something like "Initiate deprovisioning by cordoning the node(s) so no further pods can be scheduled on the node."

Author: Sounds good. I decided to remove the `to-be-deprovisioned`. LMK what you think.

4. Pre-spin any replacement nodes needed as calculated in Step (2), and wait for them to become ready.
   1. If a replacement node fails to initialize, un-cordon the to-be-deprovisioned node(s), and restart from at Step (1).
5. Delete the to-be-deprovisioned node(s) and wait for the termination controller to gracefully shutdown the node(s).
6. Once the termination controller terminates the node, go back to Step (1), starting at the the first method again.

### Termination Controller

> When a Karpenter node is deleted, the Karpenter finalizer will block deletion and the APIServer will set the `DeletionTimestamp` on the node, allowing Karpenter to gracefully shutdown the node, modeled after [K8s Graceful Node Shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown). Karpenter's graceful shutdown process will:

Reviewer: In some cases you squish `APIServer` together and other place it has a space. It doesn't matter to me either way but can we be consistent throughout the docs?

1. Cordon the node, so no further pods can be scheduled on the node.
2. Begin evicting the pods on the node with the [K8s Eviction API](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/) to respect PDBs, while ignoring all non-daemonset pods and [static pods](https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/).

> 3. While draining the node in Step (2):

Reviewer: Is this really its own step? This doesn't read to me as a step but as an aside for step 2

    # Delete all nodes owned by a specific provisioner
    kubectl delete nodes -l karpenter.sh/provisioner-name=$PROVISIONER_NAME
    ```
* **Provisioner Deletion**: Nodes are owned by the Provisioner through an [owner reference](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications) that launched them. Karpenter will gracefully terminate nodes through cascading deletion when the owning provisioner is deleted.

### Automated Methods

> * **Emptiness**: Karpenter notes when the last workload (non-daemonset) pod stops running on a node. From that point, Karpenter waits the number of seconds set by `ttlSecondsAfterEmpty` in the provisioner, then Karpenter requests to delete the node. This feature can keep costs down by removing nodes that are no longer being used for workloads.

Author: synced offline, going to do this in a follow up

* **Consolidation**: Karpenter works to actively reduce cluster cost by identifying when nodes can be removed as their workloads will run on other nodes in the cluster and when nodes can be replaced with cheaper variants due to a change in the workloads.
* **Drift**: Karpenter will annotate nodes as drifted and deprovision nodes that have drifted from their desired specification. Currently, Karpenter will only automatically mark nodes as drifted in the case of a drifted AMI.
* **Interruption**: If enabled, Karpenter will watch for upcoming involuntary interruption events that could affect your nodes (health events, spot interruption, etc.) and will cordon, drain, and terminate the node(s) ahead of the event to reduce workload disruption.

> {{% alert title="Note" color="primary" %}}

Reviewer: Not for this PR but a comment: This note should probably be broken up when we create individual sections for each of these things in a subsequent PR

    # Delete all nodes owned by a specific provisioner
    kubectl delete nodes -l karpenter.sh/provisioner-name=$PROVISIONER_NAME
    ```

> * **Provisioner Deletion**: Nodes are considered to be "owned" by the Provisioner that launched them. Karpenter will gracefully terminate nodes when a provisioner is deleted.

Reviewer: ```suggestion
* **Provisioner Deletion**: Nodes are owned by the Provisioner through an [owner reference](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications) that launched them. Karpenter will gracefully terminate nodes through cascading deletion when the owning provisioner is deleted.
```

"owned" shouldn't be in quotes. This _is_ actually what's happening.

Author: Oh I didn't write this, but I can change it :)

### Automated Methods

> * **Emptiness**: Karpenter notes when the last workload (non-daemonset) pod stops running on a node. From that point, Karpenter waits the number of seconds set by `ttlSecondsAfterEmpty` in the provisioner, then Karpenter requests to delete the node. This feature can keep costs down by removing nodes that are no longer being used for workloads.

Reviewer: We might be able to move things like use-cases into their own sections and just briefly mention the general idea of each of these methods here

* **Expiration**: Karpenter will annotate nodes as expired and deprovision nodes after they have lived a set number of seconds, based on the provisioner `ttlSecondsUntilExpired` value. One use case for node expiry is to periodically recycle nodes. Old nodes (with a potentially outdated Kubernetes version or operating system) are deleted, and replaced with nodes on the current version (assuming that you requested the latest version, rather than a specific version).
* **Consolidation**: Karpenter works to actively reduce cluster cost by identifying when nodes can be removed as their workloads will run on other nodes in the cluster and when nodes can be replaced with cheaper variants due to a change in the workloads.
* **Drift**: Karpenter will annotate nodes as drifted and deprovision nodes that have drifted from their desired specification. Currently, Karpenter will only automatically mark nodes as drifted in the case of a drifted AMI.

> * **Interruption (AWS ONLY)**: If enabled, Karpenter will watch for upcoming involuntary interruption events that could affect your nodes (health events, spot interruption, etc.) and will cordon, drain, and terminate the node(s) ahead of the event to reduce workload disruption.

Author: Ok, I'll remove 👍

### Automated Methods
* **Emptiness**: Karpenter notes when the last workload (non-daemonset) pod stops running on a node. From that point, Karpenter waits the number of seconds set by `ttlSecondsAfterEmpty` in the provisioner, then Karpenter requests to delete the node. This feature can keep costs down by removing nodes that are no longer being used for workloads.
* **Expiration**: Karpenter will annotate nodes as expired and deprovision nodes after they have lived a set number of seconds, based on the provisioner `ttlSecondsUntilExpired` value. One use case for node expiry is to periodically recycle nodes. Old nodes (with a potentially outdated Kubernetes version or operating system) are deleted, and replaced with nodes on the current version (assuming that you requested the latest version, rather than a specific version).
* **Consolidation**: Karpenter works to actively reduce cluster cost by identifying when:

>   * nodes can be removed as their workloads will run on other nodes in the cluster.

Reviewer: Any reason these two lines aren't capitalized at the beginning?
